core/src/main/java/com/datastrato/gravitino/lock/LockManager.java

```diff
@@ -0,0 +1,138 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.Stack;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * LockManager is a lock manager that manages the lock of the resource path. It will lock the whole
+ * path from root to the resource path.
+ *
+ * <p>Assuming we need to alter the table `metalake.catalog.db1.table1`, the lock manager will lock
+ * the following
+ *
+ * <pre>
+ *   /                                    readLock
+ *   /metalake                            readLock
+ *   /metalake/catalog                    readLock
+ *   /metalake/catalog/db1                readLock
+ *   /metalake/catalog/db/table1          writeLock
+ * </pre>
+ *
+ * If the lock manager fails to lock the resource path, it will release all the locks that have been
+ * locked.
+ */
+public class LockManager {
+  private static final Logger LOG = LoggerFactory.getLogger(LockManager.class);
+  private static final long TIME_AFTER_LAST_ACCESS_TO_EVICT_IN_MS = 24 * 60 * 60 * 1000; // 1 day
+
+  @VisibleForTesting final TreeLockNode treeLockRootNode = new TreeLockNode("root");
+  private final ThreadLocal<Stack<TreeLockPair>> holdingLocks = ThreadLocal.withInitial(Stack::new);
+  private final ScheduledThreadPoolExecutor lockCleaner;
+
+  static class TreeLockPair {
+    private final TreeLockNode treeLockNode;
+    private final LockType lockType;
+
+    public TreeLockPair(TreeLockNode treeLockNode, LockType lockType) {
+      this.treeLockNode = treeLockNode;
+      this.lockType = lockType;
+    }
+
+    public static TreeLockPair of(TreeLockNode treeLockNode, LockType lockType) {
+      return new TreeLockPair(treeLockNode, lockType);
+    }
+  }
+
+  public LockManager() {
+    this.lockCleaner =
+        new ScheduledThreadPoolExecutor(
+            1,
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setNameFormat("tree-lock-cleaner-%d")
+                .build());
+
+    lockCleaner.scheduleAtFixedRate(
+        () ->
+            treeLockRootNode
+                .getAllChildren()
+                .forEach(
+                    child ->
+                        evictStaleNodes(
+                            TIME_AFTER_LAST_ACCESS_TO_EVICT_IN_MS, child, treeLockRootNode)),
+        0,
+        10,
+        TimeUnit.MINUTES);
+  }
+
+  /**
+   * Evict the stale nodes from the tree lock node if the last access time is earlier than the given
+   * time.
+   *
+   * @param timeAfterLastAccessToEvictInMs The time after last access to evict the node.
+   * @param treeNode The tree lock node to evict.
+   * @param parent The parent of the tree lock node.
+   */
+  @VisibleForTesting
+  void evictStaleNodes(
+      long timeAfterLastAccessToEvictInMs, TreeLockNode treeNode, TreeLockNode parent) {
```

Comment:
What's the breaking point of this recursive method. From my understanding, we should scan from leaf nodes.

Reply:
I understand your meaning, but I believe scanning from the root may be more efficient. 

If a node has not been visited for a long time, we can freely remove it and all its descendants.  However, if we scan from the leaf node, we can only remove it and can not determine its parent's status. Additionally,  reading from root node is more space-intensive than from the root. 

Another point, regardless of the sequence we use, I believe a recursive method is necessary.